post.md

```diff
@@ -0,0 +1,411 @@
+storage. We are usually separated from all other devices but
+we use OS interface, which reuses concepts we use for
+those two. Generally speaking, everything that happens in 
+a computer is series of passing data between components and
+transforming them in between.
+
+For example, when we display a web page we can think of it:
+
+1. we have given an address in memory.
+2. transform it into request conforming standard.
+3. pass a request to the networking device
+4. accept response
+5. transform text into an image
+6. pass the image to graphics card
+
+The job of application is just to transform the data and
+inform OS where is the product, what it is and where to pass
+it. RAM is a medium for that exchange as well as storage for
+middle products. RAM also store instructions how to perform
+those transformations. Assembly language is a textual
+representation of codes understood by CPU.
+
+# Memory
+
+We can think of RAM as a function. Every byte in memory has
+its ordinal number. We use it to read or change its value.
+Bytes are usually accessed in groups of 4(32bit) or 8(64-bit).
+
+X86 CPU never use two memory locations at the same time. That's
+why it has own memory called registers. In modern x86
+architecture, there are 16 64-bit general purpose registers
+called: RAX, RBX, RCX, RDX, RDI (destination index), RSI
+(source index), RBP (base pointer), RSP (stack pointer),
+R8-R15. Despite meaningful names of some, only RSP preserved special
+meaning (it was important when it was usual to code directly
+in assembly). "R" letter in beginning denote its 64-bits as
+during 40 years of x86 evolution registers grown from 16-bit
+(AX, BX, etc), to 32-bit (EAX, EBX, etc). Among special
+registers, we have RIP (instruction pointer) stores pointer
+(= ordinal number in memory) of next instruction to run and
+EFLAGS which register special situations (like zeroing
+register or overflow while adding). Those special registers
+are never accessed directly.
+
+To move data between registers and memory is called `MOV`.
+Example:
+
+```assembly
+	mov $0xff, %rax
+```
+
+which load constant value 0xff to RAX register. This form of
+assembly language is called AT&T and mainly widespread in
+GNU world. Other popular is Intel Syntax. The most important
+difference is argument order and lack of sigils:
+
+```assembly
+	mov rax, 0xff
+```
+
+In this document, I use AT&T syntax. Other variants:
+
+```assembly
+	mov %rax, %rbx    # RAX -> RBX
+	mov %rbx, (%rax)  # RBX -> memory pointed in RAX
+	mov 3(%rcx), %rax # RAX <- memory pointed in RCX+3
+	movb $0xde, (%rdx) # 0xde-> memory pointed in RDX
+	movq $0xde, (%rdx) # same, but zeroes other 7 bytes
+	mov %ax, my_val   # 16-bit from AX to labeled memory
+	mov my_val, %ebx  # 32-bits from my_val to EBX
+	move 0xfffe, (%rax, %rbx) # 0xfffe - RAX+RBX mem
+```
+
+Note that labels in machine code are just constants. Labels
+are for programmers convenience. Remembering addresses for
```

Comment:
Note that, labels in machine code are just constants. Labels are for programmers' convenience.